lib/incoming_mail/handler.rb

```diff
@@ -60,53 +65,32 @@ def forward_msg(msg)
     end
 
     def create_comment(msg)
-      proposal = find_proposal(find_public_id(msg)) or return
-      comment_text = find_comment_text(msg)
-      comment_user = find_comment_user(msg)
       # IMPORTANT that we check/add as observer before we create comment,
       # since comment will create as a user if not already,
       # and we want the reason logged.
-      unless proposal.has_subscriber?(comment_user)
-        add_user_as_observer(comment_user, proposal)
-      end
-      comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
-      Dispatcher.on_comment_created(comment) # sends email
-      comment
-    end
+      parsed_email = InboundMailParser.new(msg)
+      proposal = parsed_email.proposal
+      user = parsed_email.comment_user
 
-    def add_user_as_observer(user, proposal)
-      observer_role = Role.find_or_create_by(name: 'observer')
-      observation = Observation.new(user_id: user.id, role_id: observer_role.id, proposal_id: proposal.id)
-      proposal.observations << observation
-      Dispatcher.on_observer_added(observation, "Added comment via email reply")
-      observation
-    end
-
-    def find_proposal(public_id)
-      Proposal.find_by_public_id(public_id) || Proposal.find(public_id)
-    end
+      return unless user  # cannot create comment for non-existent user
```

Comment:
My line numbers might not match up with your comment. Feel free to refactor as you see fit. The important thing is to avoid all the comment/observer creation if we do not have a valid User in play.